compilers/concrete-optimizer/concrete-optimizer/src/dag/unparametrized.rs

```diff
@@ -392,6 +423,43 @@ impl<'dag> DagBuilder<'dag> {
     }
 }
 
+#[derive(Clone, PartialEq, Debug, Default)]
+pub(crate) struct CompositionRules(HashMap<OperatorIndex, Vec<OperatorIndex>>);
+
+impl CompositionRules {
+    pub(crate) fn add(&mut self, from: OperatorIndex, to: OperatorIndex) {
+        let _ = self
+            .0
+            .entry(to)
+            .and_modify(|e| e.push(from))
+            .or_insert_with(|| [from].into());
+    }
+
+    pub(crate) fn update_index(&mut self, old_to_new_map: &[usize]) {
+        let old_keys = self.0.keys().map(|n| n.0).collect::<Vec<usize>>();
```

Comment:
it looks like it remove all keys and values. why not generating a new hashmap ?
it's simpler and clearer.
```rust
let new_map = ... ;
for (key, key_vect) in map.into_iter() {
    new_map.insert(
        OperatorIndex(old_to_new_map[key.0]),
        key_vect.map(from| old_to_new_map[from.0]).collect()
    )
}
self.0 = new_map;
}
```

Reply:
Indeed this is much better ! Thanks !